Adopt B95 SoC

liucoldstar opened the pull request:
Adaptation for B95 SoC with correct HAL drivers and libs

s07641069 commented on drivers/gpio/gpio_b9x.c:
Look like needs to be reworked. This a common file for B9X SoCs

liucoldstar replied:
Done.

s07641069 commented on soc/riscv/riscv-privilege/telink_b9x/soc.c:
remove space changes

liucoldstar replied:
Will do.

damien0x0023 commented:
Comment Test (no actual meaning)

s07641069 commented on drivers/gpio/gpio_b9x.c:
```diff
@@ -325,11 +369,19 @@ static void gpio_b9x_up_down_res_set(volatile struct gpio_b9x_t *gpio,
 
 	pin = BIT(pin);
 	val = up_down_res & 0x03;
+	#if CONFIG_SOC_RISCV_TELINK_B92
```
do we need to add B91 SoC here as well?

haiwentelink replied:
For B91 don't have PROT(F),so will not add

s07641069 replied:
checking if port available for soc is done by macros IS_PORT_F
Here is my proposal to clean up this part of code:
```
	#if CONFIG_SOC_RISCV_TELINK_B91 || CONFIG_SOC_RISCV_TELINK_B92
	if (IS_PORT_F(gpio)) {
		analog_reg = 0x23 + ((pin & 0xf0) ? 1 : 0);
	} else {
		analog_reg = 0x0e + (GET_PORT_NUM(gpio) << 1) + ((pin & 0xf0) ? 1 : 0);
	}
	#elif CONFIG_SOC_RISCV_TELINK_B95
	if ((IS_PORT_F(gpio)) || (IS_PORT_G(gpio))) {
		return;
	} else {
		analog_reg = 0x17 + (GET_PORT_NUM(gpio) << 1) + ((pin & 0xf0) ? 1 : 0);
	}
	#endif
```

haiwentelink replied:
yes , I have checked it will be OK , and the code will be clean .

s07641069 commented on drivers/gpio/gpio_b9x.c:
```diff
@@ -809,6 +873,18 @@ static void gpio_b9x_irq_connect_4(void)
 #endif
 
+/* If instance 4 is present and has interrupt enabled, connect IRQ */
+#if DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) > 5
```
do we need this part of code for all SoCs or only for B95?

s07641069 replied:
if there no impact into B91&B92 we can leave as it is, just double check

haiwentelink replied:
will add B95 define for the new add part .

haiwentelink replied:
I look into the code ,and find it's changed by Dmytro , for B95 , the io set high and low will use different register .

#if CONFIG_SOC_RISCV_TELINK_B91 || CONFIG_SOC_RISCV_TELINK_B92
#define GPIO_SET_LOW_LEVEL(gpio, pin)   WRITE_BIT(gpio->output, pin, 0)
#define GPIO_SET_HIGH_LEVEL(gpio, pin)  WRITE_BIT(gpio->output, pin, 1)
#elif CONFIG_SOC_RISCV_TELINK_B95
#define GPIO_SET_LOW_LEVEL(gpio, pin)   WRITE_BIT(gpio->output_clr, pin, 1)
#define GPIO_SET_HIGH_LEVEL(gpio, pin)  WRITE_BIT(gpio->output, pin, 1)
#endif
 and the code GPIO_SET_LOW_LEVEL(gpio, pin); is equal to the WRITE_BIT(gpio->output, pin, 0);
also the WRITE_BIT(gpio->output, pin, 0); is equal to gpio->output &= ~BIT(pin);

#define WRITE_BIT(var, bit, set) \
	((var) = (set) ? ((var) | BIT(bit)) : ((var) & ~BIT(bit)))
so if we want to write gpio , we should use the GPIO_SET_LOW_LEVEL or GPIO_SET_HIGH_LEVEL, can not directly operate WRITE_BIT or directly use register .

s07641069 commented on soc/riscv/riscv-privilege/telink_b9x/soc.c:
```diff
@@ -294,7 +358,7 @@ static int soc_b9x_check_flash(void)
 SYS_INIT(soc_b9x_init, PRE_KERNEL_1, 0);
-
-SYS_INIT(soc_b9x_check_flash, POST_KERNEL, 0);
+/* SYS_INIT(soc_b9x_check_flash, POST_KERNEL, 0); */
```
same here, disable only for b95 soc

s07641069 commented:
Create PR for HAL if there are changes.
And in case if specific HAL needed - change hal revision in west.yml into pull link instead of hash for PR verification

andriy-bilynskyy commented:
The following west manifest projects have been modified in this Pull Request:

| Name | Old Revision | New Revision | Diff |
| ---- | ------------ | ------------ |------|
| hal_telink | https://github.com/telink-semi/hal_telink/commit/0f03ecc893bd5818338ebccf860474310099905e (develop) | https://github.com/telink-semi/hal_telink/pull/77 | [telink-semi/hal_telink#77/files](https://github.com/telink-semi/hal_telink/pull/77/files) |

*Note: This message is automatically posted and updated by the Manifest GitHub Action.*

s07641069 commented:
Still minor changes required described above.
Try to fix all CI jobs for B91 and B92, this will make SoC dependency more clear.

haiwentelink commented:
For the driver update ,the clk rst and clk en will do in the sys.c , so don't need to do the rst and en in the drvier .

haiwentelink committed:
riscv: B95: update pwm and uart driver

- clean the clk rst and clk en for pwm .
- clean the clk rst and clk en for uart .

Signed-off-by: Haiwen Xia <haiwen.xia@telink-semi.com>

charlie-ai committed:
riscv: telink: update i2c driver

- update DTS of i2c in tlsr9258a-pinctrl.dtsi and tlsr9258a-common.dtsi
- include the sht3xd example in the build list

Signed-off-by: Lingzhi Chen <lingzhi.chen@telink-semi.com>

charlie-ai committed:
riscv: telink: update pinctrl driver

- update DTS of uart in tlsr9258a-pinctrl.dtsi
- fix gpio_function in pinctrl_b9x.c

Signed-off-by: Lingzhi Chen <lingzhi.chen@telink-semi.com>

charlie-ai committed:
riscv: telink: update watchdog

- add the watchdog clock setting

Signed-off-by: Lingzhi Chen <lingzhi.chen@telink-semi.com>

Commit:
riscv: B95: update rf driver

- update ieee802154_b9x driver and ieee802154_hw_tx security driver
- update overlayconf file of cli/coprocessor samples
- modify reboot register in soc.c

Signed-off-by: Zhenghuan Zhang <zhenghuan.zhang@telink-semi.com>

haiwentelink committed:
riscv: B95: update b95 uart driver

- open uart setting
- change the dtsi setting.
- update uart_b9x setting.

Signed-off-by: Haiwen Xia <haiwen.xia@telink-semi.com>

andy-liu-telink committed:
telink_b95: enable hwinfo

- enable hwinfo

Signed-off-by: Andy Liu <jie.liu@telink-semi.com>

s07641069 commented on .github/workflows/telink-b91-build.yaml:
what is the reason to modify B91 yaml file?

haiwentelink replied:
will revert B91

s07641069 commented on .github/workflows/telink-b92-build.yaml:
Same here
what is the reason to modify B92 yaml file?

haiwentelink replied:
will revert B92

s07641069 commented on .github/workflows/telink-b95-build.yaml:
in case if B91 and B92 yaml files were modified by mistake - just revert B91 & B 92 yaml file and will be good to make B95 as mach as similar to B91/B92 yaml files

haiwentelink replied:
change back the B91,B92 , but the B95 not support so many samples , we can only keep the sample we already tested , and will add the other samples after we add new code but we use the same style as the B92

s07641069 commented on drivers/gpio/gpio_b9x.c:
```diff
@@ -440,9 +490,9 @@ static int gpio_b9x_pin_configure(const struct device *dev,
 	/* Set GPIO init state if defined to avoid glitches */
 	if ((flags & GPIO_OUTPUT_INIT_HIGH) != 0) {
-		gpio->output |= BIT(pin);
+		GPIO_SET_HIGH_LEVEL(gpio, pin);
 	} else if ((flags & GPIO_OUTPUT_INIT_LOW) != 0) {
-		gpio->output &= ~BIT(pin);
+		GPIO_SET_LOW_LEVEL(gpio, pin);
 	}
```
double check this change, not sure if they are correct

haiwentelink replied:
already changed

s07641069 replied:
somehow I don't see any change in latest commit
have you made some change or you wants to leave as it is?

haiwentelink replied:
I look into the code ,and find it's changed by Dmytro , for B95 , the io set high and low will use different register .

```
#if CONFIG_SOC_RISCV_TELINK_B91 || CONFIG_SOC_RISCV_TELINK_B92
#define GPIO_SET_LOW_LEVEL(gpio, pin)   WRITE_BIT(gpio->output, pin, 0)
#define GPIO_SET_HIGH_LEVEL(gpio, pin)  WRITE_BIT(gpio->output, pin, 1)
#elif CONFIG_SOC_RISCV_TELINK_B95
#define GPIO_SET_LOW_LEVEL(gpio, pin)   WRITE_BIT(gpio->output_clr, pin, 1)
#define GPIO_SET_HIGH_LEVEL(gpio, pin)  WRITE_BIT(gpio->output, pin, 1)
#endif
 and the code GPIO_SET_LOW_LEVEL(gpio, pin); is equal to the WRITE_BIT(gpio->output, pin, 0);
also the WRITE_BIT(gpio->output, pin, 0); is equal to gpio->output &= ~BIT(pin);

#define WRITE_BIT(var, bit, set) \
	((var) = (set) ? ((var) | BIT(bit)) : ((var) & ~BIT(bit)))
```
      so if we want to write gpio , we should use the GPIO_SET_LOW_LEVEL or GPIO_SET_HIGH_LEVEL, can not directly operate WRITE_BIT or directly use register .

haiwentelink replied:
already test sample/basic/blinky and sample/basic/button on B91 ,and also measured the voltage, i think there will be no problem .

s07641069 commented on drivers/hwinfo/hwinfo_b9x.c:
```diff
@@ -13,9 +13,13 @@ ssize_t z_impl_hwinfo_get_device_id(uint8_t *buffer, size_t length)
 	flash_mid = flash_read_mid();
-
 	flash_read_mid_uid_with_check((unsigned int *)&flash_mid, uid);
+#else
```
use #elif CONFIG_SOC_RISCV_TELINK_B95

s07641069 commented on drivers/ieee802154/ieee802154_b9x.c:
```diff
@@ -1042,8 +1045,6 @@ static int b9x_stop(const struct device *dev)
 		riscv_plic_irq_disable(DT_INST_IRQN(0) - CONFIG_2ND_LVL_ISR_TBL_OFFSET);
 		rf_set_tx_rx_off();
-		rf_baseband_reset();
-		rf_reset_dma();
```
what is the reason to remove?
As I remember this code required for B91 & B92

haiwentelink replied:
For the initial version not have these interface ,have already add .

s07641069 commented on drivers/ieee802154/ieee802154_b9x_frame.c:
```diff
@@ -612,7 +618,12 @@ ieee802154_b9x_crypto_ecb(
+#ifdef CONFIG_SOC_RISCV_TELINK_B95
+	(void)ske_lp_crypto(SKE_ALG_AES_128, SKE_MODE_ECB,
+	SKE_CRYPTO_ENCRYPT, (uint8_t *)key, 0, NULL, (uint8_t *)inp, out, 16);
+#else
```
lets use everywhere one style as above:
#if CONFIG_SOC_RISCV_TELINK_B91 || CONFIG_SOC_RISCV_TELINK_B92
.....
#elif CONFIG_SOC_RISCV_TELINK_B95
....
#endif

s07641069 replied:
added another comment, looks like #else needs to be removed now

s07641069 commented on drivers/serial/uart_b9x.c:
```diff
@@ -187,9 +187,13 @@ static void uart_b9x_cal_div_and_bwpc(uint32_t baudrate, uint32_t pclk,
 static void uart_b9x_init(volatile struct uart_b9x_t *uart, uint16_t divider,
 			  uint8_t bwpc, uint8_t parity, uint8_t stop_bit)
 {
+#if CONFIG_SOC_RISCV_TELINK_B95
+	uart->ctrl0 = ((uart->ctrl0 & (~FLD_UART_BPWC_O)) | bwpc);
+#else
+	uart->ctrl0 = bwpc;
+#endif
```
same here lets use one style for #if SoCs as above
and what is the reason to move this upper line "divider = ..." ?

haiwentelink replied:
actually it's meaning less , but i think if need ,we shoud move the code like that(related code works together)
/* config clock */
	divider = divider | FLD_UART_CLK_DIV_EN;
        uart->clk_div = divider;

s07641069 commented on samples/net/openthread/cli/boards/tlsr9258a.overlay:
```diff
@@ -0,0 +1,11 @@
+/*
+ * Copyright (c) 2023 Telink Semiconductor
```
2024

s07641069 commented on samples/net/openthread/coprocessor/boards/tlsr9258a.conf:
```diff
@@ -0,0 +1,22 @@
+# Telink B92 TLSR9258A
```
B95

s07641069 commented on soc/riscv/riscv-privilege/telink_b9x/soc.c:
```diff
@@ -8,7 +8,9 @@
 #include <clock.h>
 #include <gpio.h>
 #include <ext_driver/ext_pm.h>
+/*
 #include "rf.h"
+*/
```
why this include is disabled?

s07641069 replied:
use #if for B95 only if this is temporary needed

s07641069 commented on soc/riscv/riscv-privilege/telink_b9x/soc.c:
```diff
@@ -280,14 +354,16 @@ static int soc_b9x_check_flash(void)
 	}
 
 	if (hw_flash_size < dts_flash_size) {
+		/**
 		printk("!!! flash error: expected (.dts) %u, actually %u\n",
 			dts_flash_size, hw_flash_size);
+		**/
```
why this error print is disabled?

s07641069 commented on drivers/ieee802154/ieee802154_b9x_frame.c:
```diff
@@ -603,7 +603,15 @@ b9x_ieee802154_get_data(const uint8_t *payload,
+#elif CONFIG_SOC_RISCV_TELINK_B95
+#include <ske.h>
+#include <ske_portable.h>
+#else
```
we haven't other SoCs in B9X family, why we need here #else  ?
looks like can be removed

haiwentelink replied:
OK, i will check with zhenghuan , seems can be removed.

haiwentelink replied:
already update code

s07641069 commented on soc/riscv/riscv-privilege/telink_b9x/Kconfig.soc:
```diff
@@ -38,7 +38,7 @@ config TELINK_B9x_HWDSP
 	bool "Use Telink B9x platform mbedtls HW acceleration"
 	depends on SOC_SERIES_RISCV_TELINK_B9X && MBEDTLS
-	default y
+	default n
```
This will disable HW crypto for all B9X SoCs
Revert this change and if you need to disable only for B95, jut set "CONFIG_TELINK_B9X_MBEDTLS_HW_ACCELERATION=n" in some conf or Kconfig file of B95 SoC (board)

s07641069 replied:
TELINK_B9X_MBEDTLS_HW_ACCELERATION=y required for Matter
there two cases:
1. Disable only for B95 and will need to rework Matter. Anyway will need in near future.
2. Revert to Y and create jira ticket for B95 to Adopt/fix and test HW crypto for B95. Better to do this.

haiwentelink replied:
already revert in Kconfig.soc , and set the TELINK_B9X_MBEDTLS_HW_ACCELERATION=y in the tlsr9258a_defconfig

pulin1103 replied:
About CONFIG_TELINK_B9X_MBEDTLS_HW_ACCELERATION, I have create a ticket ZEPHYR-503, see my comment in it.

s07641069 commented on soc/riscv/riscv-privilege/telink_b9x/soc.c:
```diff
@@ -20,7 +20,11 @@
 #endif
 
 /* Software reset defines */
+#ifdef CONFIG_SOC_RISCV_TELINK_B95
+#define reg_reset                   REG_ADDR8(0x14082f)
+#else
 #define reg_reset                   REG_ADDR8(0x1401ef)
+#endif
```
Again use same style:
```
#if CONFIG_SOC_RISCV_TELINK_B91 || CONFIG_SOC_RISCV_TELINK_B92
	...
#elif CONFIG_SOC_RISCV_TELINK_B95
	...
#endif
```

haiwentelink replied:
OK, will check all the define of the CONFIG_SOC_RISCV_TELINK_B95, to make sure all the code to follow the same style .

s07641069 commented on soc/riscv/riscv-privilege/telink_b9x/soc.c:
```diff
@@ -58,14 +64,14 @@
-		#error "Wrong value for power-mode parameter"
-	#endif
+	#error "Wrong value for power-mode parameter"
+#endif
```
check tabs, for this #endif it is missed

haiwentelink replied:
fixed

s07641069 commented on soc/riscv/riscv-privilege/telink_b9x/soc.c:
```diff
@@ -226,32 +261,69 @@ void soc_b9x_restore(void)
 
+static inline uint32_t read_flash_mid(void)
+{
+#if CONFIG_SOC_RISCV_TELINK_B91 || CONFIG_SOC_RISCV_TELINK_B92
```
again one style
```
#if CONFIG_SOC_RISCV_TELINK_B91 || CONFIG_SOC_RISCV_TELINK_B92
	...
#elif CONFIG_SOC_RISCV_TELINK_B95
	...
#endif
```

haiwentelink replied:
fixed

andy-liu-telink committed:
telink_b95: enable flash capacity check at boot

- enable flash capacity check
- tidy code up
- update the hal_telink revision
- archive liblt_9258_zephyr.a for testing
- enable compilation of blinky sample

Signed-off-by: Andy Liu <jie.liu@telink-semi.com>

andy-liu-telink committed:
telink_b95: add building hello_world example

- include the hello_world example in the build list

Signed-off-by: Andy Liu <jie.liu@telink-semi.com>

haiwentelink committed:
riscv: B95: update b95 pwm driver

- add the pwm clock setting and pwm start proc
- update pin set , for the value exceed .
- update west.yml for telink .

Signed-off-by: Haiwen Xia <haiwen.xia@telink-semi.com>

charlie-ai committed:
riscv: B95: update watchdog driver & dts

- clean the clk rst and clk en for watchdog
- update DTS of pwm in tlsr9258a-common.dtsi
- Update comments in tlsr9528a-pinctrl.dtsi

Signed-off-by: Lingzhi Chen <lingzhi.chen@telink-semi.com>

andy-liu-telink committed:
telink_b95: enable simplified blinky example

- enable 32kHz RC clock initialization at boot
- update dts files to support B95
- disable some code to ensure successful compilation
- update mtimer & plic addresses
- add code to verify mtimer and plic

Signed-off-by: Andy Liu <jie.liu@telink-semi.com>

andy-liu-telink committed:
telink_b95: enable gpio driver

- modify dts files
- update gpio_b9x and pinctrl_b9x
- restore the original code in blinky/main.c

Signed-off-by: Andy Liu <jie.liu@telink-semi.com>

haiwentelink committed:
riscv: telink: update flash & 4-wire mode

- open the flash check part.
- certify the flash driver.
- update 4-wire setting.
- test the 2-wire and 4-wire mode.

Signed-off-by: Haiwen Xia <haiwen.xia@telink-semi.com>

charlie-ai committed:
riscv: telink: update gpio driver

- rename gpio register
- update pin function
- update DTS of key in tlsr9258a-common.dtsi
- fix io_pullup_down issues in gpio_b9x.c

Signed-off-by: Lingzhi Chen <lingzhi.chen@telink-semi.com>

haiwentelink committed:
riscv: B95: update module and ci

- change the west.yml to hal_telink
- add peripheral_ht in bluetooth
- choice IEEE802154_B9X_MAC_TYPE
- clean code for gpio ,uart hwinfo driver.
- clean code for the soc.c for the header file and print part.
- update the setting and code for 802154.
- revert the setting for TELINK_B9X_MBEDTLS_HW_ACCELERATION.
- clean code for the code style.

Signed-off-by: Haiwen Xia <haiwen.xia@telink-semi.com>
